File: spine/orreth_spine/inbox.py
# PROVENANCE: Claude Fable 5 (claude-fable-5) — rearch P1 sp1, the durability boundary (M1) · 2026-09-16
"""The durable inbox (canon 0002, laws 3 and 4).

Delivery is at-least-once; EFFECTS are once. Every side-effecting
consumer records the message id it acted on in the same transaction as
the effect, so a redelivered message finds its own footprint and steps
around it. Aggregate ordering is enforced per aggregate, never globally:
an older sequence is recorded and skipped, the next sequence applies,
and a GAP refuses to guess — the caller fetches what's missing instead
of applying history out of order.
"""
from __future__ import annotations


class GapDetected(Exception):
    """A sequence arrived from the future: expected `expected`, got `got`
    for `aggregate_id`. The consumer must recover the missing events —
    applying out of order would let an older truth overwrite a newer one."""

    def __init__(self, aggregate_id: str, expected: int, got: int):
        self.aggregate_id, self.expected, self.got = aggregate_id, expected, got
        super().__init__(
            f"gap on {aggregate_id}: expected sequence {expected}, got {got}"
            " — fetch the missing events; never guess")


def ensure_schema(conn) -> None:
    from .outbox import once
    if not once(conn, "inbox"):
        return
    with conn.transaction():
        cur = conn.cursor()
        cur.execute("SELECT pg_advisory_xact_lock(742199)")  # DDL race guard
        cur.execute(
            "CREATE TABLE IF NOT EXISTS spine_inbox ("
            " consumer text NOT NULL,"
            " message_id text NOT NULL,"
            " first_seen timestamptz NOT NULL DEFAULT now(),"
            " status text NOT NULL DEFAULT 'working',"
            " attempts int NOT NULL DEFAULT 1,"
            " PRIMARY KEY (consumer, message_id))")
        cur.execute(
            "CREATE TABLE IF NOT EXISTS spine_aggregate_cursor ("
            " consumer text NOT NULL,"
            " aggregate_id text NOT NULL,"
            " last_sequence bigint NOT NULL DEFAULT 0,"
            " PRIMARY KEY (consumer, aggregate_id))")


def apply_once(conn, consumer: str, message_id: str, effect) -> str:
    """Run `effect(cur)` exactly once per (consumer, message id). The
    inbox footprint and the effect share one transaction: if the effect
    fails, the footprint rolls back too and a redelivery retries clean.
    Returns "applied" or "duplicate"."""
    with conn.transaction():
        cur = conn.cursor()
        cur.execute(
            "INSERT INTO spine_inbox (consumer, message_id)"
            " VALUES (%s, %s) ON CONFLICT (consumer, message_id) DO NOTHING"
            " RETURNING message_id", (consumer, message_id))
        if cur.fetchone() is None:
            cur.execute(
                "UPDATE spine_inbox SET attempts = attempts + 1"
                " WHERE consumer = %s AND message_id = %s",
                (consumer, message_id))
            return "duplicate"
        effect(cur)
        cur.execute(
            "UPDATE spine_inbox SET status = 'done'"
            " WHERE consumer = %s AND message_id = %s", (consumer, message_id))
    return "applied"


def apply_event(conn, consumer: str, env: dict, effect) -> str:
    """apply_once plus the ordering law for envelopes wearing an
    aggregate {id, sequence}: stale sequences are recorded and skipped,
    the next sequence applies and advances the cursor, and a gap raises
    GapDetected inside a rolled-back transaction. Returns "applied",
    "duplicate", or "stale"."""
    agg = env.get("aggregate") or {}
    aid = str(agg.get("id") or "")
    seq = int(agg.get("sequence") or 0)
    if not aid or seq <= 0:
        return apply_once(conn, consumer, env["message_id"], effect)
    with conn.transaction():
        cur = conn.cursor()
        cur.execute(
            "INSERT INTO spine_aggregate_cursor (consumer, aggregate_id)"
            " VALUES (%s, %s) ON CONFLICT DO NOTHING", (consumer, aid))
        cur.execute(
            "SELECT last_sequence FROM spine_aggregate_cursor"
            " WHERE consumer = %s AND aggregate_id = %s FOR UPDATE",
            (consumer, aid))
        last = cur.fetchone()[0]
        if seq <= last:
            cur.execute(
                "INSERT INTO spine_inbox (consumer, message_id, status)"
                " VALUES (%s, %s, 'stale')"
                " ON CONFLICT (consumer, message_id)"
                " DO UPDATE SET attempts = spine_inbox.attempts + 1",
                (consumer, env["message_id"]))
            return "stale"
        if seq > last + 1:
            raise GapDetected(aid, last + 1, seq)
        cur.execute(
            "INSERT INTO spine_inbox (consumer, message_id)"
            " VALUES (%s, %s) ON CONFLICT (consumer, message_id) DO NOTHING"
            " RETURNING message_id", (consumer, env["message_id"]))
        if cur.fetchone() is None:
            cur.execute(
                "UPDATE spine_inbox SET attempts = attempts + 1"
                " WHERE consumer = %s AND message_id = %s",
                (consumer, env["message_id"]))
            return "duplicate"
        effect(cur)
        cur.execute(
            "UPDATE spine_aggregate_cursor SET last_sequence = %s"
            " WHERE consumer = %s AND aggregate_id = %s", (seq, consumer, aid))
        cur.execute(
            "UPDATE spine_inbox SET status = 'done'"
            " WHERE consumer = %s AND message_id = %s",
            (consumer, env["message_id"]))
    return "applied"


def duplicates_seen(conn, consumer: str) -> int:
    """The honest duplicate meter: redeliveries absorbed without effect."""
    cur = conn.cursor()
    cur.execute(
        "SELECT coalesce(sum(attempts - 1), 0) FROM spine_inbox"
        " WHERE consumer = %s", (consumer,))
    return int(cur.fetchone()[0])

File: backend/plane/crates/orreth-spine/src/inbox.rs
// PROVENANCE: Claude Fable 5.1 (claude-fable-5-1) — rearch P7 sp2, the ground and the rails · 2026-09-22
//! The durable inbox — mirrors `orreth_spine.inbox` (canon 0002, laws 3 and
//! 4). Delivery is at-least-once; EFFECTS are once. A consumer records the
//! message id it acted on in the same transaction as the effect, so a
//! redelivery finds its own footprint and steps around it. Ordering is per
//! aggregate, never global: a stale sequence is recorded and skipped, the
//! next applies, and a GAP refuses to guess.
//!
//! The tables are the Python spine's `spine_inbox` and
//! `spine_aggregate_cursor`, word for word.

use crate::ground::Ground;
use crate::rail_error::RailError;
use crate::rails::{inbox_route, InboxRoute};
use serde_json::Value;
use std::fmt;
use tokio_postgres::Transaction;

pub const DDL: &[&str] = &[
    "CREATE TABLE IF NOT EXISTS spine_inbox ( consumer text NOT NULL, message_id text NOT NULL, \
     first_seen timestamptz NOT NULL DEFAULT now(), status text NOT NULL DEFAULT 'working', \
     attempts int NOT NULL DEFAULT 1, PRIMARY KEY (consumer, message_id))",
    "CREATE TABLE IF NOT EXISTS spine_aggregate_cursor ( consumer text NOT NULL, aggregate_id \
     text NOT NULL, last_sequence bigint NOT NULL DEFAULT 0, PRIMARY KEY (consumer, \
     aggregate_id))",
];

pub async fn ensure_schema(g: &mut Ground) -> Result<(), RailError> {
    g.ensure("inbox", DDL).await.map(|_| ())
}

/// What the inbox did with a delivery.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum Outcome {
    Applied,
    Duplicate,
    Stale,
}

impl fmt::Display for Outcome {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        f.write_str(match self {
            Outcome::Applied => "applied",
            Outcome::Duplicate => "duplicate",
            Outcome::Stale => "stale",
        })
    }
}

/// The footprint: the row claimed → `true`; a duplicate bumps attempts → `false`.
async fn claim(tx: &Transaction<'_>, consumer: &str, message_id: &str) -> Result<bool, RailError> {
    let claimed = tx
        .query_opt(
            "INSERT INTO spine_inbox (consumer, message_id) VALUES ($1, $2) ON CONFLICT \
             (consumer, message_id) DO NOTHING RETURNING message_id",
            &[&consumer, &message_id],
        )
        .await?;
    if claimed.is_none() {
        tx.execute(
            "UPDATE spine_inbox SET attempts = attempts + 1 WHERE consumer = $1 AND message_id = \
             $2",
            &[&consumer, &message_id],
        )
        .await?;
        return Ok(false);
    }
    Ok(true)
}

async fn done(tx: &Transaction<'_>, consumer: &str, message_id: &str) -> Result<(), RailError> {
    tx.execute(
        "UPDATE spine_inbox SET status = 'done' WHERE consumer = $1 AND message_id = $2",
        &[&consumer, &message_id],
    )
    .await?;
    Ok(())
}

/// Run `effect` exactly once per (consumer, message id). The footprint and
/// the effect share one transaction: if the effect fails, the footprint
/// rolls back too and a redelivery retries clean.
pub async fn apply_once<F>(
    g: &mut Ground,
    consumer: &str,
    message_id: &str,
    effect: F,
) -> Result<Outcome, RailError>
where
    F: for<'t, 'c> AsyncFnOnce(&'t Transaction<'c>) -> Result<(), RailError>,
{
    let tx = g.client_mut().transaction().await?;
    if !claim(&tx, consumer, message_id).await? {
        tx.commit().await?;
        return Ok(Outcome::Duplicate);
    }
    effect(&tx).await?;
    done(&tx, consumer, message_id).await?;
    tx.commit().await?;
    Ok(Outcome::Applied)
}

/// `apply_once` plus the ordering law for envelopes wearing an aggregate
/// `{id, sequence}`: stale sequences are recorded and skipped, the next
/// sequence applies and advances the cursor, and a gap refuses inside a
/// rolled-back transaction (`RailError::Gap`).
pub async fn apply_event<F>(
    g: &mut Ground,
    consumer: &str,
    env: &Value,
    effect: F,
) -> Result<Outcome, RailError>
where
    F: for<'t, 'c> AsyncFnOnce(&'t Transaction<'c>) -> Result<(), RailError>,
{
    let (message_id, aggregate_id, sequence) = match inbox_route(env) {
        InboxRoute::Once { message_id } => {
            return apply_once(g, consumer, &message_id, effect).await;
        }
        InboxRoute::Sequenced {
            message_id,
            aggregate_id,
            sequence,
        } => (message_id, aggregate_id, sequence),
    };
    let tx = g.client_mut().transaction().await?;
    tx.execute(
        "INSERT INTO spine_aggregate_cursor (consumer, aggregate_id) VALUES ($1, $2) ON \
         CONFLICT DO NOTHING",
        &[&consumer, &aggregate_id],
    )
    .await?;
    let last: i64 = tx
        .query_one(
            "SELECT last_sequence FROM spine_aggregate_cursor WHERE consumer = $1 AND \
             aggregate_id = $2 FOR UPDATE",
            &[&consumer, &aggregate_id],
        )
        .await?
        .get(0);
    if sequence <= last {
        tx.execute(
            "INSERT INTO spine_inbox (consumer, message_id, status) VALUES ($1, $2, 'stale') ON \
             CONFLICT (consumer, message_id) DO UPDATE SET attempts = spine_inbox.attempts + 1",
            &[&consumer, &message_id],
        )
        .await?;
        tx.commit().await?;
        return Ok(Outcome::Stale);
    }
    if sequence > last + 1 {
        drop(tx); // rolled back: nothing recorded, the caller fetches what is missing
        return Err(RailError::Gap {
            aggregate_id,
            expected: last + 1,
            got: sequence,
        });
    }
    if !claim(&tx, consumer, &message_id).await? {
        tx.commit().await?;
        return Ok(Outcome::Duplicate);
    }
    effect(&tx).await?;
    tx.execute(
        "UPDATE spine_aggregate_cursor SET last_sequence = $1 WHERE consumer = $2 AND \
         aggregate_id = $3",
        &[&sequence, &consumer, &aggregate_id],
    )
    .await?;
    done(&tx, consumer, &message_id).await?;
    tx.commit().await?;
    Ok(Outcome::Applied)
}

/// The honest duplicate meter: redeliveries absorbed without effect.
pub async fn duplicates_seen(g: &Ground, consumer: &str) -> Result<i64, RailError> {
    let row = g
        .client()
        .query_one(
            "SELECT coalesce(sum(attempts - 1), 0)::bigint FROM spine_inbox WHERE consumer = $1",
            &[&consumer],
        )
        .await?;
    Ok(row.get(0))
}
